core/src/test/java/org/elasticsearch/gateway/GatewayModuleTests.java

```diff
@@ -0,0 +1,101 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.gateway;
+
+import org.elasticsearch.cluster.ClusterName;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.env.NodeEnvironment;
+import org.elasticsearch.node.Node;
+import org.elasticsearch.node.NodeBuilder;
+import org.elasticsearch.node.internal.InternalSettingsPreparer;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.InternalTestCluster;
+
+import static org.hamcrest.Matchers.instanceOf;
+
+public class GatewayModuleTests extends ESTestCase {
+
+    public void testCustomGateway() {
+        Node node = buildNode("mock");
+        try {
+            assertThat(node.injector().getInstance(Gateway.class), instanceOf(MockGateway.class));
+        } finally {
+            node.close();
+        }
+    }
+
+    public void testDefaultGateway() {
+        Node node = buildNode(null);
+        try {
+            assertThat(node.injector().getInstance(Gateway.class), instanceOf(Gateway.class));
+        } finally {
+            node.close();
+        }
+    }
+
+    protected Node buildNode(String gateway) {
+        final Settings.Builder settings = Settings.builder()
+                .put(ClusterName.SETTING, InternalTestCluster.clusterName("custom-gateway", randomLong()))
+                .put("path.home", createTempDir())
+                .put("path.shared_data", createTempDir().getParent())
+                .put("node.name", "mock_gateway")
+                .put(EsExecutors.PROCESSORS, 1) // limit the number of threads created
+                .put("http.enabled", false)
+                .put(InternalSettingsPreparer.IGNORE_SYSTEM_PROPERTIES_SETTING, true);
+        if (gateway != null) {
+            settings.put(GatewayModule.GATEWAY_TYPE_KEY, gateway);
+            settings.putArray("plugin.types", TestPlugin.class.getName());
+        }
+        return NodeBuilder.nodeBuilder().local(true).data(true).settings(settings // make sure we get what we set :)
+        ).build().start();
+    }
+
+    public static class TestPlugin extends Plugin {
+        @Override
+        public String name() {
+            return "mock-gateway";
+        }
+
+        @Override
+        public String description() {
+            return "injecting custom MockGateway testing";
+        }
+
+        public void onModule(GatewayModule gatewayModule) {
+            gatewayModule.addGatewayType("mock", MockGateway.class);
+        }
+
+        @Override
+        public Settings additionalSettings() {
+            return Settings.builder().put(GatewayModule.GATEWAY_TYPE_KEY, "mock").build();
```

Comment:
didnt follow the latest guice changes too closely (so this might be valid), but is this needed, when you configure it implicetely in ``buildNode()` above?